### ISSUE TITLE
[ARCHETYPE-159] mvn archetype:generate is failing to download needed archetype artifacts

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-archetype#569]
**[Glen Mazza](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gmazza)** opened **[ARCHETYPE-159](https://issues.apache.org/jira/browse/ARCHETYPE-159?redirect=false)** and commented

A fresh install of Maven 2.0.9 on Windows XP, if I issue this command:

mvn archetype:generate -DarchetypeArtifactId=maven-archetype-quickstart

I get this error message:
[INFO] The desired archetype does not exist (org.apache.maven.archetypes:maven-archetype-quickstart:1.0)

Partial error stack:
[INFO] Scanning for projects...
[INFO] Searching repository for plugin with prefix: 'archetype'.
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Default Project
[INFO]    task-segment: [archetype:generate] (aggregator-style)
[INFO] ------------------------------------------------------------------------
[INFO] Preparing archetype:generate
[INFO] No goals needed for project - skipping
[INFO] Setting property: classpath.resource.loader.class => 'org.codehaus.plexus
.velocity.ContextClassLoaderResourceLoader'.
[INFO] Setting property: velocimacro.messages.on => 'false'.
[INFO] Setting property: resource.loader => 'classpath'.
[INFO] Setting property: resource.manager.logwhenfound => 'false'.
[INFO] [archetype:generate]
org.apache.maven.archetype.downloader.DownloadNotFoundException: Requested downl
oad does not exist.
at org.apache.maven.archetype.downloader.DefaultDownloader.download(Defa
ultDownloader.java:62)
at org.apache.maven.archetype.common.DefaultArchetypeArtifactManager.exi
sts(DefaultArchetypeArtifactManager.java:310)
at org.apache.maven.archetype.ui.DefaultArchetypeGenerationConfigurator.
configureArchetype(DefaultArchetypeGenerationConfigurator.java:103)

The same error also happens with an archetype artifact ID of "maven-archetype-webapp" (and probably all the others as well).

However, if I issue the same command but using archetype:create instead of :generate, not only will the archetype artifact correctly download (and the command work successfully), subsequent usages of archetype:generate with that same archetype artifact will also correctly run.

So I think the problem is that mvn archetype:generate needs to download archetype artifacts just like mvn archetype:create does.

---

**Affects:** 2.0-alpha-2

**Attachments:**
- [ARCHETYPE-159-3.patch](https://issues.apache.org/jira/secure/attachment/12713848/ARCHETYPE-159-3.patch) (_5.13 kB_)

7 votes, 2 watchers

**Comment by jira-importer** [comment]
**[Olivier Martin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olivier.martin)** commented

Same error here, there is an incompatibility between the archetype plugin and Maven 2.0.9.

I can reproduce it with any of the default catalog and with our own catalog.
The only workaround is to have a local version of the archetype, it's not downloaded from the server.

```
Downloading: http://static.appfuse.org/releases/org/appfuse/archetypes/appfuse-basic-jsf/2.0/appfuse-basic-jsf-2.0.jar
org.apache.maven.archetype.downloader.DownloadException: Error downloading.
        at org.apache.maven.archetype.downloader.DefaultDownloader.download(DefaultDownloader.java:58)
        at org.apache.maven.archetype.common.DefaultArchetypeArtifactManager.exists(DefaultArchetypeArtifactManager.java:310)
        at org.apache.maven.archetype.ui.DefaultArchetypeGenerationConfigurator.configureArchetype(DefaultArchetypeGenerationConfigurator.java:103)
```

For me this is a blocker.

**Comment by jira-importer** [comment]
**[Arnaud Heritier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aheritier)** commented

Same thing here :

```
[INFO] Setting property: classpath.resource.loader.class => 'org.codehaus.plexus.velocity.ContextClassLoaderResourceLoader'.
[INFO] Setting property: velocimacro.messages.on => 'false'.
[INFO] Setting property: resource.loader => 'classpath'.
[INFO] Setting property: resource.manager.logwhenfound => 'false'.
[INFO] [archetype:generate]
Choose archetype:
1: remote -> socle-fondations-services-archetype (Couche de services)
2: remote -> socle-ia-archetype (WebApp IA (Struts))
3: remote -> socle-ria-archetype (WebApp RIA (JSF/RichFaces))
4: remote -> socle-fondations-ws-archetype (WebApp WS (XFire))
5: remote -> socle-fondations-ear-archetype (EAR)
Choose a number:  (1/2/3/4/5): 1
Downloading: file://E:\Dev\data\maven-2.x\repository/fr/XXX/socle/fondations/archetypes/socle-fondations-services-archetype/2.0.0/socle-fondations-services
-archetype-2.0.0.jar
org.apache.maven.archetype.downloader.DownloadNotFoundException: Requested download does not exist.
        at org.apache.maven.archetype.downloader.DefaultDownloader.download(DefaultDownloader.java:75)
        at org.apache.maven.archetype.common.DefaultArchetypeArtifactManager.exists(DefaultArchetypeArtifactManager.java:310)
        at org.apache.maven.archetype.ui.DefaultArchetypeGenerationConfigurator.configureArchetype(DefaultArchetypeGenerationConfigurator.java:125)
        at org.apache.maven.archetype.mojos.CreateProjectFromArchetypeMojo.execute(CreateProjectFromArchetypeMojo.java:163)
        at org.apache.maven.plugin.DefaultPluginManager.executeMojo(DefaultPluginManager.java:451)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.executeGoals(DefaultLifecycleExecutor.java:558)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.executeStandaloneGoal(DefaultLifecycleExecutor.java:512)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.executeGoal(DefaultLifecycleExecutor.java:482)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.executeGoalAndHandleFailures(DefaultLifecycleExecutor.java:330)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.executeTaskSegments(DefaultLifecycleExecutor.java:227)
        at org.apache.maven.lifecycle.DefaultLifecycleExecutor.execute(DefaultLifecycleExecutor.java:142)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:336)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:129)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:287)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:585)
        at org.codehaus.classworlds.Launcher.launchEnhanced(Launcher.java:315)
        at org.codehaus.classworlds.Launcher.launch(Launcher.java:255)
        at org.codehaus.classworlds.Launcher.mainWithExitCode(Launcher.java:430)
        at org.codehaus.classworlds.Launcher.main(Launcher.java:375)
Caused by: org.apache.maven.artifact.resolver.ArtifactNotFoundException: File: E:\Dev\data\maven-2.x\repository\fr\XXX\socle\fondations\archetypes\socle-fo
ndations-services-archetype\2.0.0\socle-fondations-services-archetype-2.0.0.jar does not exist
```

It tries to download the archetype from ..... my local repository :-(

**Comment by jira-importer** [comment]
**[Arnaud Heritier](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aheritier)** commented

It's a first patch for this bug.
There's no test, I didn't yet find how to add one for this problem.
I changed 2 things to fix the problem :
- I passed remoteRepositories list from the plugin to the Downloader
- I removed a suspect code on the the Downloader. I don't understand why you change the repository of the artifact. I you uncomment this, the bug comes back. Same thing with repositories list.
